packages/aic/src/document.ts

```diff
@@ -0,0 +1,318 @@
+  candidateDiff: Diff | Document | null, // null when only updating tid
+  stored_tick: Tick, // null when there is not db entry (did init)
+  key: Asymmetric, // the consortium passes in the key
+): Promise<Tick | ErrorMessage> => {
+  if (candidateDiff === null) {
+    // nothing to update return the old tick but with new tid
+    return (await sign(
+      {
+        tid: tid,
+        did: stored_tick.did,
+        diffs: stored_tick.diffs,
+        key: stored_tick.key,
+        sig: '',
+      },
+      key,
+    )) as Tick
+  }
+  // there is a validated stored_tick and a candidateDiff
+  // this is a update of the did document
+  if (!(await validateSig(candidateDiff, key))) {
+    return { error: 'diff has bad signature' }
+  }
+  if (
+      'key' in candidateDiff && 
+      typeof candidateDiff.key == 'string' && 
+      !validateKeyInDocForDiffs(candidateDiff.key, stored_tick.diffs)
+    ) {
+      return { error: 'diff signed by key not in did doc' }
+    }
+  // there is a validated stored_tick and a valid candidateDiff
+  const diffs: Diffs = stored_tick.diffs
+  diffs[tid] = candidateDiff
+  return tickFromDiffs(diffs, tid, key)
+}
+
+const validateKeyInDocForDiffs = async (
+  key: DidKeyString,
+  diffs: Diffs,
+) => {
+  return true // @TODO write this
+  // the 72 hour key prioraty window gos here :)
+}
+
+export const tickFromDiffs = async (
+  diffs: Diffs,
+  tid: TidString,
+  key: Asymmetric,
+): Promise<Tick | ErrorMessage> => {
+  // This function dose not validate the diffs remember to do that first
+  const tids = Object.keys(diffs).sort()
+  if (tids.length < 1) {
+    return { error: 'no diffs' }
+  }
+  const initialState = diffs[tids[0]]
+  return (await sign(
+    {
+      tid: tid,
+      did: 'did:aic:' + (await pid(initialState)),
+      diffs,
+      key: key.did(),
+      sig: '', // sig will be filld in by call to sign
+    },
+    key,
+  )) as Tick
+}
+
+const authoriseKey = (
+  doc: Document,
+  diff: Diff,
+): boolean => {
+  // @TODO the 72 hour window logic needs to be here
+  if (keyIn(doc, 'adx/account_keys', diff.key)) {
+    return true
+  }
+  if (keyIn(doc, 'adx/recovery_keys', diff.key)) {
+    return patchesKeysOnly(diff.patches)
+  }
+  return false
+}
+
+const keyIn = (doc: Document, keyList: string, didKey: DidKeyString): boolean => {
+  return keyList in doc &&
+  Array.isArray(doc[keyList]) &&
+  (doc[keyList] as String[]).includes(didKey)
+}
+
+const patchesKeysOnly = (patches: Patch[]): boolean => {
+  // check that the diff.patches[*][2]
+  // the recovey key is only authorised to update account or recovery keys
+  // only changes 'adx/account_keys' or 'adx/recovery_keys'
+  return patches.every((patch: Patch) => {
+    const pathSegment = patch[1][0]
+    return ( pathSegment === 'adx/account_keys' || pathSegment === 'adx/recovery_keys' )
+  })
+}
+
+export const tickToDidDoc = async (tick: Tick, consortiumDid: string, key: Asymmetric, asOf?: TidString,): Promise<Document> => {
```

Comment:
Reading this function signature (& others in this file), it gives me the impression that it _always_ returns a Document. Either need to return `Promise<Document | Error>` or just throw the error.

Coming from Go, I always wanted to return errors. But I know use throw/catch much more often. It prevents you from screwing up error checking (cause the error will always bubble up & not get swallowed)

Reply:
With the throw pattern we would still need to build an error JSON object at the end to return to the client.
I was using throw for errors for the server to deal with and a JSON Document for the Errors that go to the client.

What is the pattern you like to keep these clear what we are doing?